### PR TITLE
add WebAppOnLinux support

### PR DESCRIPTION
**Pull request by Eskibear**
Classname would be refactored later.

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/handlers/PublishHandler.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+		}
+
+		PublishWizard pw = new PublishWizard();
+		WizardDialog pwd = new AzureWizardDialog(window.getShell(), pw);
+		if (pwd.open() == Window.OK) {
+			ConsoleLogger.info(String.format("URL: http://%s.azurewebsites.net/%s",
+					DockerRuntime.getInstance().getLatestImageName(), project.getName()));
+		}
+		props = DockerRuntime.getInstance().saveToProps(props);
+		ConfigFileUtil.saveConfig(project, props);
```
Comment (adashen) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/handlers/PublishHandler.java]: ok to save config

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.core/src/com/microsoft/azuretools/core/components/AzureWizardPage.java [hunk 2]
```diff
@@ -8,7 +9,7 @@
 import org.eclipse.jface.resource.ImageDescriptor;
 import org.eclipse.jface.wizard.WizardPage;
 
-public abstract class AzureWizardPage extends WizardPage {
+public abstract class AzureWizardPage extends WizardPage implements MvpView{
```
Comment (adashen) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.core/src/com/microsoft/azuretools/core/components/AzureWizardPage.java]: any impact to the existing sub classes of AzureWizardPage?

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/presenters/StepOnePagePresenter.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+import rx.Observable;
+import rx.schedulers.Schedulers;
+
+import com.microsoft.azuretools.container.ConsoleLogger;
+import com.microsoft.azuretools.container.DockerRuntime;
+
+public class StepOnePagePresenter<V extends StepOnePage> extends MvpPresenter<V> {
+
+    // StepOne Page
+    public boolean onValidateRegistry(String registryUrl, String registryUsername, String registryPassword) {
```
Comment (adashen) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/presenters/StepOnePagePresenter.java]: change name to pushImage?

Comment (adashen) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/presenters/StepOnePagePresenter.java]: Limit it to ACR?

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/handlers/PublishHandler.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+		} catch (DockerCertificateException | DockerException | InterruptedException | IOException e) {
+			e.printStackTrace();
+			return null;
+		}
+
+		PublishWizard pw = new PublishWizard();
+		WizardDialog pwd = new AzureWizardDialog(window.getShell(), pw);
+		if (pwd.open() == Window.OK) {
+			ConsoleLogger.info(String.format("URL: http://%s.azurewebsites.net/%s",
+					DockerRuntime.getInstance().getLatestImageName(), project.getName()));
```
Comment (Eskibear) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/handlers/PublishHandler.java]: DockerRuntime.getInstance().getLatestWebAppName()

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/presenters/StepTwoPagePresenter.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+            AzureModelController.updateSubscriptionMaps(null);
+            return WebAppOnLinuxUtil.listAllWebAppOnLinux();
+        })
+        .subscribeOn(Schedulers.io())
+        .subscribe(wal -> {
+            DefaultLoader.getIdeHelper().invokeLater(() -> {
+                getMvpView().finishLoading(wal);
+            });
+        }, e -> {
+            ConsoleLogger.error("onLoadWebAppsOnLinux@StepTwoPagePresenter");
```
Comment (adashen) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/presenters/StepTwoPagePresenter.java]: should consolelogger called from ui thread?

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/PublishWizard.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package com.microsoft.azuretools.container.ui.wizard.publish;
+
+import org.eclipse.jface.wizard.Wizard;
+
+public class PublishWizard extends Wizard {
```
Comment (adashen) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/PublishWizard.java]: should it extend from azurewizard?

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/StepOnePage.java [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+import org.eclipse.swt.layout.GridData;
+import org.eclipse.swt.widgets.Text;
+import org.eclipse.ui.forms.widgets.FormToolkit;
+import org.eclipse.swt.widgets.Display;
+import org.eclipse.swt.custom.StyledText;
+
+import org.eclipse.swt.custom.ScrolledComposite;
+import org.eclipse.swt.widgets.Button;
+
+public class StepOnePage extends AzureWizardPage {
```
Comment (adashen) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/StepOnePage.java]: does telemetry added for these pages?

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/StepOnePage.java [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        setPageComplete(false);
+    }
+
+    private void onBtnValidateSelection() {
+        setWidgetsEnabledStatus(false);
+        presenter.onValidateRegistry(txtRegistryUrl.getText(), txtUsername.getText(), txtPassword.getText());
+    }
+
+    @Override
+    protected void finalize() throws Throwable {
```
Comment (adashen) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/StepOnePage.java]: do not release resources from finalize. the method may not be called

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/StepTwoPage.java [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+
+	private void onBtnCreateSelection() {
+		StepTwoPopupDialog dialog = new StepTwoPopupDialog(this.getShell());
+		if (dialog.open() == Window.OK) {
+			presenter.onLoadWebAppsOnLinux();
+		}
+	}
+
+	@Override
+	protected void finalize() throws Throwable {
```
Comment (adashen) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/StepTwoPage.java]: move to dispose

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/StepOnePage.java [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+    private Text txtRegistryUrl;
+    private Text txtUsername;
+    private Text txtPassword;
+    private StyledText styledText;
+    private final FormToolkit formToolkit = new FormToolkit(Display.getDefault());
+    private final StepOnePagePresenter<StepOnePage> presenter;
+    private Button btnValidate;
+
+    // Call Presenter
+    public boolean validated() {
```
Comment (adashen) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/StepOnePage.java]: not used?

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/StepOnePage.java [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+        return presenter.onValidateRegistry(txtRegistryUrl.getText(), txtUsername.getText(), txtPassword.getText());
+    }
+
+    public void updateRegistryInfo() {
```
Comment (adashen) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/StepOnePage.java]: not used?

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/StepTwoPopupDialog.java [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+		presenter.onLoadSubsAndRGs();
+	}
+
+	private void radioResourceGroupLogic() {
+		boolean enabled = btnResourceGroupCreateNew.getSelection();
+		textResourceGroupName.setEnabled(enabled);
+		comboResourceGroup.setEnabled(!enabled);
+	}
+
+	protected void finalize() throws Throwable {
```
Comment (adashen) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/StepTwoPopupDialog.java]: move to dispose

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/utils/WebAppOnLinuxUtil.java [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+				DockerRuntime.getInstance().getRegistryUsername(), DockerRuntime.getInstance().getRegistryPassword());
+		String imageName = DockerRuntime.getInstance().getLatestImageName();
+		return deploy(subscriptionId, resourceGroup, appName, pr, imageName, createNewRg);
+	}
+
+	public static WebApp deploy(String subscriptionId, String resourceGroup, String appName, PrivateRegistry pr,
+			String imageName, boolean createNewRg) throws IOException {
+		AzureManager azureManager = AuthMethodManager.getInstance().getAzureManager();
+		// not signed in
+		if (azureManager == null) {
```
Comment (adashen) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/utils/WebAppOnLinuxUtil.java]: ask user to signin?

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/handlers/PublishHandler.java (hunk 4 above)
Comment (Eskibear) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/handlers/PublishHandler.java]: ack

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/presenters/StepOnePagePresenter.java (hunk 3 above)
Comment (Eskibear) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/presenters/StepOnePagePresenter.java]: ack

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.core/src/com/microsoft/azuretools/core/components/AzureWizardPage.java (hunk 2 above)
Comment (Eskibear) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.core/src/com/microsoft/azuretools/core/components/AzureWizardPage.java]: no for the moment.

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/handlers/PublishHandler.java (hunk 1 above)
Comment (Eskibear) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/handlers/PublishHandler.java]: ack

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/presenters/StepTwoPopupDialogPresenter.java [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+import com.microsoft.azuretools.container.utils.WebAppOnLinuxUtil;
+import com.microsoft.azuretools.utils.AzureModel;
+import com.microsoft.tooling.msservices.components.DefaultLoader;
+
+import rx.Observable;
+import rx.schedulers.Schedulers;
+
+public class StepTwoPopupDialogPresenter<V extends StepTwoPopupDialog> extends MvpPresenter<V> {
+	private List<SubscriptionDetail> binderSubscriptionDetails;
+	private List<ResourceGroup> binderResourceGroup;
```
Comment (Eskibear) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/presenters/StepTwoPopupDialogPresenter.java]: ack

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/presenters/StepTwoPopupDialogPresenter.java [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+import com.microsoft.azuretools.container.ui.wizard.publish.StepTwoPopupDialog;
+import com.microsoft.azuretools.container.utils.WebAppOnLinuxUtil;
+import com.microsoft.azuretools.utils.AzureModel;
+import com.microsoft.tooling.msservices.components.DefaultLoader;
+
+import rx.Observable;
+import rx.schedulers.Schedulers;
+
+public class StepTwoPopupDialogPresenter<V extends StepTwoPopupDialog> extends MvpPresenter<V> {
+	private List<SubscriptionDetail> binderSubscriptionDetails;
```
Comment (Eskibear) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/presenters/StepTwoPopupDialogPresenter.java]: ack

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/PublishWizard.java (hunk 6 above)
Comment (Eskibear) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/PublishWizard.java]: there's no `AzureWizard` but `AzureWizardDialog`, refer to related handler to see current behaviour.

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/StepOnePage.java (hunk 8 above)
Comment (Eskibear) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/StepOnePage.java]: ack

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/StepTwoPage.java (hunk 9 above)
Comment (Eskibear) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/StepTwoPage.java]: ack

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/StepOnePage.java (hunk 10 above)
Comment (Eskibear) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/StepOnePage.java]: removed

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/StepTwoPopupDialog.java (hunk 12 above)
Comment (Eskibear) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/StepTwoPopupDialog.java]: ack

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/utils/WebAppOnLinuxUtil.java (hunk 13 above)
Comment (Eskibear) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/utils/WebAppOnLinuxUtil.java]: ack, will require sign-in in handler

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/utils/ConfigFileUtil.java [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+        return prop;
+    }
+    
+    public static void saveConfig(IProject project, Properties prop){
+        IFile file = project.getFile(configFileName);
+        
+        try {
+            ByteArrayOutputStream ous = new ByteArrayOutputStream();
+            prop.store(ous, null);
+            file.create(new ByteArrayInputStream(ous.toByteArray()), true, null);
```
Comment (Eskibear) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/utils/ConfigFileUtil.java]: ack

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/presenters/StepTwoPagePresenter.java (hunk 5 above)
Comment (Eskibear) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/presenters/StepTwoPagePresenter.java]: It should only update dataStream in ConsoleLogger without UI events. will refine ConsoleLogger

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/presenters/StepTwoPopupDialogPresenter.java (hunk 15 above)
Comment (Eskibear) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/presenters/StepTwoPopupDialogPresenter.java]: ack

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/presenters/StepTwoPopupDialogPresenter.java (hunk 14 above)
Comment (Eskibear) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/presenters/StepTwoPopupDialogPresenter.java]: ack

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/StepOnePage.java (hunk 7 above)
Comment (Eskibear) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/ui/wizard/publish/StepOnePage.java]: added

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.core/src/com/microsoft/azuretools/core/components/AzureWizardPage.java (hunk 2 above)
Comment (Eskibear) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.core/src/com/microsoft/azuretools/core/components/AzureWizardPage.java]: But I'd better remove it

**Pull request merged**